### ISSUE TITLE
The authorization headers in your request are not valid: Nonce already used

### ISSUE THREAD

**Issue by cariaso** [mwclient/mwclient#164]
I've recently installed https://www.mediawiki.org/wiki/Extension:Moderation and it's doing it's job well, but it seems to be interfering with mwclient. I'd appreciate your input in trying to isolate the cause as either mwclient or https://www.mediawiki.org/wiki/Extension:Moderation as the source of the trouble.

My bot runs fine for ~30 minutes (that's an underinformed estimate, and it's not yet clear if that number is consistent between runs, and certainly something I can determine if it will help to diagnose), but then it crashes. The most recent crash is during a read, not a write, but it's possible all API calls are affected. The traceback is

```python

Traceback (most recent call last):
  File "./orientalbot.py", line 1344, in <module>
    main()
  File "./orientalbot.py", line 1302, in main
    oriental.tasks.clinvar.dualload()
  File "/Users/cariaso/dev/rrb2016/rrb/snpedia/bots/oriental/tasks/clinvar.py", line 816, in dualload
    dualSetClinVar()
  File "/Users/cariaso/dev/rrb2016/rrb/snpedia/bots/oriental/tasks/clinvar.py", line 808, in dualSetClinVar
    dualstore(snpedia, name,rsnum,snp)
  File "/Users/cariaso/dev/rrb2016/rrb/snpedia/bots/oriental/tasks/clinvar.py", line 740, in dualstore
    process(snpedia, rsnum, mem[rsnum])
  File "/Users/cariaso/dev/rrb2016/rrb/snpedia/bots/oriental/tasks/clinvar.py", line 692, in process
    apage = snpedia.mwbot.pages[rsnum]
  File "/usr/local/lib/python2.7/site-packages/mwclient/listing.py", line 224, in __getitem__
    return self.get(name, None)
  File "/usr/local/lib/python2.7/site-packages/mwclient/listing.py", line 255, in get
    return cls(self.site, full_page_name, info)
  File "/usr/local/lib/python2.7/site-packages/mwclient/page.py", line 35, in __init__
    inprop='protection', *extra_props)
  File "/usr/local/lib/python2.7/site-packages/mwclient/client.py", line 228, in get
    return self.api(action, 'GET', *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/mwclient/client.py", line 282, in api
    if self.handle_api_result(info, sleeper=sleeper):
  File "/usr/local/lib/python2.7/site-packages/mwclient/client.py", line 306, in handle_api_result
    info['error']['info'], info['error']['*'])
mwclient.errors.APIError: (u'mwoauth-invalid-authorization', u'The authorization headers in your request are not valid: Nonce already used: 109459792799187288961496196369', u'See https://bots.SNPedia.com/api.php for API usage')

```

I'm guessing that the mwclient is caching a nonce that needs to change periodically? But it's certainly possible that something about Moderation is the source of the issue. However so far I'm only seeing this error via the mwclient. I'm hoping that this traceback will help you to draw a conclusion, and that will help me to formulate a more specific question to the developer of 'Moderation' instead of leaving me with "It doesn't work".

FWIW, the mwclient is talking to https://bots.snpedia.com

while I've not yet tested it, it seems very likely that this can be reproduced with just

```python
self = {}
self.mwbot = mwclient.Site(('https','bots.snpedia.com'), path='/',
    consumer_token='A', consumer_secret='B',access_token='C',access_secret='D'
                     )
for rsnumpage in self.mwbot.Categories['Is_a_snp']:
    rsnum = rsnumpage.title
    apage = snpedia.mwbot.pages[rsnum]
```

where A,B,C,D are gotten from
https://bots.snpedia.com/index.php/Special:OAuthConsumerRegistration/propose

I'm happy to do whatever further testing or isolation would be useful, but perhaps what's here is enough to make a diagnosis trivial?

**Comment by danmichaelo** [comment]
Hm, that's odd. The request signing is done by request-oauthlib (which uses oauthlib for most of the work), so I'm quite sure this cannot be attributed to mwclient. See e.g. [here](https://hueniverse.com/beginners-guide-to-oauth-part-iii-security-architecture-e9394f5263b5) for an explanation of OAuth signing and nonces.

**Comment by danmichaelo** [comment]
Note: If you set logging level to DEBUG, you can have all the nonces (together with lots of other stuff) printed out:

```python 
import logging
logging.basicConfig(level=logging.DEBUG)
```

(you have to do this before importing anything else)

**Comment by cariaso** [comment]
https://phabricator.wikimedia.org/T106066
indicates that while the text of the error message is perhaps poor (in that it should propagate or better communicate the upstream error from memcached or redis), but that fundamentally this is not something that the mediawiki folks seem inclined to change on the server side. 

https://phabricator.wikimedia.org/T109173
indicates that the error is well known to the pywikibot code, and that they've addressed it via
https://gerrit.wikimedia.org/r/#/c/289582/
which is a fairly simple change
https://phabricator.wikimedia.org/rPWBC2bcdc2ecf158e1db858220c1977742cd1c83cfdc

I will probably try to make a similar change to mwclient and offer a pull request for your consideration.

**Comment by danmichaelo** [comment]
Interesting… I'm happy to accept a PR for such a a workaround then

**Comment by danmichaelo** [comment]
Fixed in #165
